dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp

```diff
@@ -3921,6 +3932,8 @@ RtpsUdpDataLink::RtpsWriter::~RtpsWriter()
       ACE_TEXT("deleting with %d elements left not fully acknowledged\n"),
       elems_not_acked_.size()));
   }
+
+  heartbeat_.cancel_and_wait();
```

Comment:
I'm not sure this method should be holding on the the mutexes if it's going to potentially be waiting here. Based on @mitza-oci 's comments in the past I'm not sure locking either one in the destructor is a great idea anyway, since it means somewhere out there someone is accessing a reference counted object without a reference count... so maybe try removing them and fix the incorrect access if it causes problems?